### PR TITLE
Ad pay tests

### PR DESCRIPTION
**Pull request by RightFS**
{Summary}

Fixes #{issue}

**Type of change**

* [ ] Bug Fix (non-breaking change which fixes an issue)
* [ ] New Feature (non-breaking change which adds functionality)
* [ ] Breaking Change (fix or feature that would cause existing functionality to not work as expected)
* [ ] This change requires a documentation update
* [ ] This change requires a storage migration
* [ ] This change requires a benchmark

**Checklist**

* [ ] My code follows the style guidelines of this project
* [ ] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas
* [ ] I have made corresponding changes to the documentation
* [ ] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [ ] New and existing unit tests pass locally with my changes
* [ ] New and existing benchmarks pass locally with my changes
* [ ] Any dependent changes have been merged and published in downstream modules
* [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: pallets/ad/src/tests.rs [hunk 1]
```diff
@@ -427,15 +431,87 @@ fn should_pay() {
             slot.tokens
         );
         assert_eq!(slot.remain, 400 - 40);
-        assert_eq!(slot.tokens, 19 - 5);
+        assert_eq!(slot.tokens, 19 - 2);
 
-        assert_eq!(Assets::balance(nft_meta.token_asset_id, &CHARLIE), 5);
+        assert_eq!(Assets::balance(nft_meta.token_asset_id, &CHARLIE), 2);
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:436]: 怎么知道CHARLIE应该获得两个碎片？从测试的上下文看不出原因。

File: pallets/ad/src/tests.rs [hunk 2]
```diff
@@ -427,15 +431,87 @@ fn should_pay() {
...
 
         assert_eq!(
             Tag::get_score(&DID_CHARLIE, vec![0u8, 1u8, 2u8, 3u8, 4u8, 5u8]),
             11
         );
 
+        // 2.2 pay2
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:443]: 写在单独的测试用例中，一个测试用例只测一件事

File: pallets/ad/src/tests.rs [hunk 3]
```diff
@@ -427,15 +431,87 @@ fn should_pay() {
...
+            slot.tokens
+        );
+
+        assert_eq!(Assets::balance(nft_meta.token_asset_id, &TAGA5_TAGB2), 3);
+
+        // 2.3 pay3
+        assert_ok!(Ad::pay(
+            Origin::signed(BOB),
+            ad,
+            nft,
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:465]: 独立测试用例

File: pallets/ad/src/tests.rs [hunk 4]
```diff
@@ -427,15 +431,87 @@ fn should_pay() {
...
+
+        // 2.4 pay4
+        assert_ok!(Ad::pay(
+            Origin::signed(BOB),
+            ad,
+            nft,
+            DID_TAGA100_TAGB100,
+            vec![(vec![0u8, 1u8, 2u8, 3u8, 4u8, 5u8], 5)],
+            None
+        ));
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:487]: 独立测试用例

File: pallets/ad/src/tests.rs [hunk 5]
```diff
@@ -427,15 +431,87 @@ fn should_pay() {
...
+        assert_eq!(
+            Assets::balance(nft_meta.token_asset_id, &meta.pot),
+            slot.tokens
+        );
+
+        assert_eq!(Assets::balance(nft_meta.token_asset_id, &TAGA100_TAGB100), 10);
+
+        // 2.5 pay5
+        assert_ok!(Ad::pay(
+            Origin::signed(BOB),
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:499]: 独立测试用例

File: pallets/ad/src/tests.rs (hunk 1 above)
Comment (RightFS) [reply pallets/ad/src/tests.rs:436]: (用户命中的 TAG 分数和) / (广告TAG数量) 然后取整.  [(5+0)/2]=2

File: pallets/ad/src/tests.rs [hunk 6]
```diff
@@ -421,35 +428,105 @@ fn should_pay() {
...
-                Origin::signed(BOB),
-                ad,
-                nft,
-                DID_CHARLIE,
-                vec![(vec![0u8, 1u8, 2u8, 3u8, 4u8, 5u8], 5)],
-                None
-            ),
-            Error::<Test>::Paid
+#[test]
+fn should_pay_1() {
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:451]:
```suggestion
fn should_pay_3_for_taga$5_tagb$2() {
```

File: pallets/ad/src/tests.rs [hunk 7]
```diff
@@ -421,35 +428,105 @@ fn should_pay() {
...
+            vec![(vec![0u8, 1u8, 2u8, 3u8, 4u8, 5u8], 5)],
+            None
+        ));
+
+        assert_eq!(Assets::balance(nft_meta.token_asset_id, &TAGA5_TAGB2), 3);
+    });
+}
+
+#[test]
+fn should_pay_3() {
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:471]:
```suggestion
fn should_pay_0_when_all_tags_score_are_zero() {
```

File: pallets/ad/src/tests.rs [hunk 8]
```diff
@@ -421,35 +428,105 @@ fn should_pay() {
...
+            DID_TAGA0_TAGB0,
+            vec![(vec![0u8, 1u8, 2u8, 3u8, 4u8, 5u8], 5)],
+            None
+        ));
+
+        assert_eq!(Assets::balance(nft_meta.token_asset_id, &TAGA0_TAGB0), 0);
+    });
+}
+#[test]
+fn should_pay_4() {
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:490]:
```suggestion
fn should_pay_10_when_all_tags_are_full_score() {
```

File: pallets/ad/src/tests.rs [hunk 9]
```diff
@@ -421,35 +428,105 @@ fn should_pay() {
...
+        ));
+
+        assert_eq!(
+            Assets::balance(nft_meta.token_asset_id, &TAGA100_TAGB100),
+            10
         );
     });
 }
+#[test]
+fn should_pay_5() {
```
Comment (coderfengyun) [inline pallets/ad/src/tests.rs:512]: should_pay_10_when_all_tags_are_fullscore_or_overflow

**Review by coderfengyun — APPROVED** [review]

**Pull request merged**